### ISSUE TITLE
[macOS] "New Mach-O header is too large to relocate" when using py2app

### ISSUE THREAD

**Issue by albertosottile** [pyca/cryptography#7847]
### Description:
When creating an application bundle (`.app`) from Python code on macOS, some tools like [py2app](https://github.com/ronaldoussoren/py2app) rewrite the library linking to make the bundle self consistent and not linked to system libraries. 

This process works well with `cryptography` up to 38.0.1. However, version 38.0.3 raises this error in `py2app`
```
ValueError: New Mach-O header is too large to relocate in '/Users/runner/work/syncplay/syncplay/dist/Syncplay.app/Contents/Resources/lib/python3.10/lib-dynload/cryptography/hazmat/bindings/_rust.so' (new size=2264, max size=2260, delta=72)
```

This means that the header space allocated in the shared library is not enough to accommodate the new (probably longer) destination path. According to the [py2app FAQ](https://py2app.readthedocs.io/en/latest/faq.html) 

> This error can be avoided by rebuilding binaries with enough space in the Mach-O headers, either by using the linker flag “-headerpad_max_install_names” or by installing shared libraries in a deeply nested location (the path for the install root needs to be at least 30 characters long).

Indeed, I built `cryptography==38.0.3` from sources with
```
env LDFLAGS="-L$(brew --prefix openssl@3)/lib -Wl -headerpad_max_install_names" CFLAGS="-I$(brew --prefix openssl@3)/include" pip install --no-binary :all: cryptography
```

and I can confirm this fixes the issue in `py2app` and allows again to create bundles that depend on `cryptography`.

Would it be possible for you to add this linker flag in your build chain?

### Impact:
This bug prevents creation of any Python-based application bundle on macOS if the bundle itself depends on `cryptography`. Current workaround is to pin the dependency on 38.0.1. 

### Environment:
macOS 12.5
Python 3.10.6
cryptography 38.0.3
cffi 1.15.1
pip 22.3.1
setuptools 65.6.3

**Comment by reaperhulk** [comment]
Fascinating. I don't have any issue with adding that flag to our wheel builders going forward, but I'm surprised 38.0.1 -> 38.0.3 caused this problem since no paths should have changed. The `_rust.abi3.so` we ship only links against CoreFoundation, libiconv, libSystem, and libresolv:

```
_rust.abi3.so (architecture x86_64):
	@rpath/_rust.abi3.so (compatibility version 0.0.0, current version 0.0.0)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1858.112.0)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.100.3)
	/usr/lib/libresolv.9.dylib (compatibility version 1.0.0, current version 1.0.0)
_rust.abi3.so (architecture arm64):
	@rpath/_rust.abi3.so (compatibility version 0.0.0, current version 0.0.0)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1858.112.0)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.100.3)
	/usr/lib/libresolv.9.dylib (compatibility version 1.0.0, current version 1.0.0)
```

Presumably it's possible that our Mach-O header increased in size for some other reason, but I'd love to know what happened just for my own edification before we land the fix. 😄

**Comment by reaperhulk** [comment]
Looking a bit more deeply, in the example you've provided, `_rust` shows as the problem, but passing flags to `LDFLAGS` and `CFLAGS` doesn't propagate to `rustc` through `cargo` (you need to use `RUSTFLAGS`). This suggests that something else in your environment (different `rustc` perhaps?) is what resolved the issue.

As an aside, the flags you've provided aren't quite correct in that `-Wl` has no meaning in the context of `LDFLAGS`. `LDFLAGS` are passed directly to the linker, while `CFLAGS` are passed to the compiler. In cases where you need to proxy linker flags through the compiler you can do `-Wl,-headerpad_max_install_names` (note the comma) though. Additionally, your example dynamically links `libssl` and `libcrypto` against the `_openssl.abi3.so`, where we statically link it in our wheels, but since the error message is for the rust object that probably doesn't matter.

It would be very helpful in debugging this to have a way to determine whether `headerpad_max_install_names` has actually done anything. We don't want to land a wrong or broken fix!

**Comment by albertosottile** [comment]
Understood. I am not sure how I can further assist you without fully duplicating your build chain (and it is quite evident that I do not fully know it...). Would you be willing to investigate this on your side if I provide the code for a minimal app bundle?

**Comment by albertosottile** [comment]
I created a minimal example here: https://gist.github.com/albertosottile/97d001ed6177ef8cad9b6680af57f84f

Usage: 
- On a macOS system, prepare a virtual environment with installed `pip`, `setuptools`, `wheel`, `cryptography`, and `py2app`.
-`python setup.py py2app`

The bundle can be created with cryptography 38.0.1 or with the x86_64 wheels of 38.0.3. When using the `universal2` wheels of 38.0.3, the issue described here occurs.
```
ValueError: New Mach-O header is too large to relocate in '/Users/alberto/cryptography_bundle/dist/hello.app/Contents/Resources/lib/python3.10/lib-dynload/cryptography/hazmat/bindings/_rust.so' (new size=2264, max size=2260, delta=72)
```

**Comment by reaperhulk** [comment]
Thanks for the minimal example, I was able to reproduce the error locally. After some debugging it appears this problem occurs when compiled using rust 1.64.0 (which is what 38.0.3 used), but not 1.63.0 (38.0.1) or 1.65.0 (current release). I don't really understand **why** though, so I'd still like to find a way to effectively dissect these headers and understand what's going on.

**Comment by albertosottile** [comment]
> it appears this problem occurs when compiled using rust 1.64.0 (which is what 38.0.3 used), but not 1.63.0 (38.0.1) or 1.65.0 (current release).

Does this mean the problem will not occur in the next version of the wheels?

> I'd still like to find a way to effectively dissect these headers and understand what's going on.

I think at this point we need the help of @ronaldoussoren.

**Comment by alex** [comment]
Yes, this looks like it'd be fixed with our next wheels (whether 38.0.4 or 39.0).

**Comment by ronaldoussoren** [comment]
> Thanks for the minimal example, I was able to reproduce the error locally. After some debugging it appears this problem occurs when compiled using rust 1.64.0 (which is what 38.0.3 used), but not 1.63.0 (38.0.1) or 1.65.0 (current release). I don't really understand **why** though, so I'd still like to find a way to effectively dissect these headers and understand what's going on.

py2app (or rather macholib) rewrites load commands in shared libraries, such as CPython extensions, to refer to paths in the generated application bundle instead of the absolute path in a shared library.   The replacement value might be longer than the original value, and that's something macholib can only do when the load command has enough room.  By default the size of the load command is just enough for the included path, for C extensions "-W-l,-headerpath_max_install_names" grows the size of those commands to the maximum size, which should be ample size for macholib to do its work.

Looking at the universal2 wheel for cryptography 38.0.3 I don't understand yet why the shared libaries cause a problem, the load commands in those should libraries shouldn't require rewriting in the first place.

**Comment by ronaldoussoren** [comment]
> Looking at the universal2 wheel for cryptography 38.0.3 I don't understand yet why the shared libaries cause a problem, the load commands in those should libraries shouldn't require rewriting in the first place.

I now do understand what's going on: I forgot that py2app/macholib doesn't properly support ``@rpath`` and will rewrite load commands with ``@rpath``.  Fixing this is on my (too long) todo list, but requires some substantial changes to my code because macholib was written before ``@rpath`` and similar constructs existed and that has resulted in some design patterns for macholib being invalid. 

The easiest fix for now is to use the header path option mentioned earlier.

BTW. To inspect the load commands: ``otool -vl .../_rust.ab3.so`` will print all load commands including their sizes.

**Comment by alex** [comment]
Ok, we've released 38.0.4 which is built with a newer rustc, which we believe fixes this.
